src/shims/panic.rs

```diff
@@ -119,19 +119,54 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
 
     fn handle_stack_pop(
         &mut self,
-        mut extra: FrameData<'tcx>,
+        mut frame: Frame<'mir, 'tcx, Tag, FrameData<'tcx>>,
         unwinding: bool,
     ) -> InterpResult<'tcx, StackPopJump> {
         let this = self.eval_context_mut();
 
-        trace!("handle_stack_pop(extra = {:?}, unwinding = {})", extra, unwinding);
+        trace!("handle_stack_pop(frame.extra = {:?}, unwinding = {})", frame.extra, unwinding);
+
+        if unwinding {
+            for &def_id in &[frame.instance.def_id(), frame.body.source.def_id()] {
+                let attrs = this.tcx.codegen_fn_attrs(def_id);
+                if !attrs.flags.contains(CodegenFnAttrFlags::UNWIND) {
+                    if attrs.flags.contains(CodegenFnAttrFlags::RUSTC_ALLOCATOR_NOUNWIND) {
+                        throw_ub_format!(
+                            "unwind across a function that is marked as `#[rustc_allocator_nounwind]`"
+                        );
+                    }
+                    // `fn_sig()` can't be used on closures, but closures always have "rust-call" ABI
+                    // which always allows unwinding anyway.
+                    if !this.tcx.is_closure(def_id) {
+                        let abi = this.tcx.fn_sig(def_id).abi();
+                        if !matches!(
```

Comment:
Please add a FIXME mentioning https://github.com/rust-lang/rust/issues/65303 -- we want to just be able to do `abi.may_unwind()` or so at some point.